lib/rucio/client/downloadclient.py

```diff
@@ -73,6 +73,134 @@ def __init__(self, _client=None, user_agent='rucio_clients', logger=None):
+            no_subdir           - Optional: If true, files are written directly into base_dir and existing files are overwritten. (Default: False)
+            ignore_checksum     - Optional: If true, the checksum validation is skipped (for pfn downloads the checksum must be given explicitly). (Default: True)
+            transfer_timeout    - Optional: Timeout time for the download protocols. (Default: None)
+        :param num_threads: Suggestion of number of threads to use for the download. It will be lowered if it's too high.
+        :param trace_custom_fields: Custom key value pairs to send with the traces
+
+        :returns: a list of dictionaries with an entry for each file, containing the input options, the did, and the clientState
+                  clientState can be one of the following: ALREADY_DONE, DONE, FILE_NOT_FOUND, FAIL_VALIDATE, FAILED
+
+        :raises InputValidationError: if one of the input items is in the wrong format
+        :raises NoFilesDownloaded: if no files could be downloaded
+        :raises NotAllFilesDownloaded: if not all files could be downloaded
+        :raises RucioException: if something unexpected went wrong during the download
+        """
+        logger = self.logger
+        trace = copy.deepcopy(self.trace_tpl)
+        log_prefix = 'Extracting'
+
+        logger.info('Processing %d item(s) for input' % len(items))
+        input_items = []
+        for item in items:
+            archive = item.get('archive')
+            file_extract = item.get('did')
+            rse_name = item.get('rse')
+            file_extract_scope, file_extract_name = self._split_did_str(file_extract)
+            archive_scope, archive_name = self._split_did_str(archive)
+
+            # if not specified, taking first availbale replica
+            if not rse_name:
+                replicas = self.client.list_replicas([{'scope': archive_scope, 'name': archive_name}],
+                                                     schemes='root',
+                                                     client_location=detect_client_location())
+                for r in replicas:
+                    for rse in r['states'].keys():
+                        if r['states'][rse] == 'AVAILABLE':
+                            rse_attr = self.client.list_rse_attributes(rse)
+                            if rse_attr['istape'] == 'False':
+                                rse_name = rse
+                                break
+                    if rse_name and rse_name is not None:
+                        break
+
+            rse = {}
+            try:
+                rse = rsemgr.get_rse_info(rse_name)
+            except RSENotFound:
+                logger.warning('%sCould not get info of RSE %s' % (log_prefix, rse_name))
+                continue
+
+                raise InputValidationError('The keys archive (zip file), file_extract, and rse are mandatory')
```

Comment:
this raise will probably not be reached

Reply:
the raise is redundant anyway, and rse is actualy optional. I removed this line.